sys/cheri/cheri_cap_relocs.c

```diff
@@ -0,0 +1,10 @@
+#include <cheri_init_globals.h>
+
+/* Invoked from locore. */
+extern void init_cap_relocs(void *data_cap, void *pc_cap);
+
+void
+init_cap_relocs(void *data_cap, void *pc_cap)
+{
+	cheri_init_globals_3(data_cap, pc_cap, data_cap);
```

Comment:
Any calls to cheri_init_globals need to be in standalone files so they can include the compiler-provided header safely, and I can't call into this file from another C file since that will use cap_relocs that aren't yet populated.  I'm not quite sure I follow though what DYNAMIC will contain that will be useful here?  It won't have existing caps to use for data_cap or pc_cap for example.  I guess it has the location of __caprelocs instead of using symbols so we could call cheri_init_globals_impl?  I think powerpc is currently the only relocatable kernel in FreeBSD.  RISC-V isn't, so I'm not sure a dynamic relocbase is as useful?

Reply:
You can include the header in whatever file you like? The dynamic table has DT_RISCV___CAPRELOCS(SZ) entries or whatever the exact name is, equivalent to the DT_RELA(SZ) used by PowerPC. You can get data_cap and pc_cap using builtins, and relocbase is effectively data_cap, but used exclusively for deriving capabilities to the ELF sections. I just think it's better to do something that looks like (a) what another architecture does (b) what RTLD does. Plus, later on, when we kill __cap_relocs and use proper ELF relocations, it'll make the required changes smaller, as there you really do need to use the dynamic tags to find the relocations. And finally calling it `elf_relocate_self` abstracts away the exact details whilst also being more consistent.

Reply:
I have to turn on all the compiler builtin headers for this file to compile it, but that would (potentially) affect the headers that, say, elf_machdep.c would use.  I isolated the this file to avoid dealing with worrying about that.  I think data_cap may not be ddc at the time I call this function anymore (I might have already cleared ddc), but I could perhaps defer clearing of ddc to later.  I do think the name is probably better.  I had hoped that having a common file would mean we could share it across architectures, but MIPS already has to do two passes (and the MIPS versions are all open-coded without using the compiler builtin in at all which is another approach we could take I guess)

Reply:
Ah for RTLD I used relocbase as data_cap but got PCC for pc_cap (code_cap...).

Reply:
We'll also need to solve the builtin headers issue soon anyway, because we should be using (or wrapping) cheriintrin.h. One option would be to vendor the headers if you're concerned there are issues, but I'd prefer to just make sure the search order is correct.

Reply:
Hmm, in general FreeBSD uses -nostdinc for the kernel build to turn off /usr/include and it also turns off the builtins.  The few places that need to use builtins (e.g. for vector instructions on x86) currently handle this by building specific files without -nostdinc.  cheriintrin.h is indeed going to be a bit of a bear.

Reply:
Can we add `-isystem ${COMPILER_RESOURCE_DIR}`?